pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java

```diff
@@ -96,8 +106,15 @@
   // likely that we get fresh data each time instead of multiple copies of roughly same data.
   private static final int MIN_INTERVAL_BETWEEN_STATS_UPDATES_MINUTES = 30;
 
+  // TODO(upsert): TableConfig is not available at class init phase, so we have to always create a new upsertMetadataTableManager
+  private UpsertMetadataTableManager _upsertMetadataTableManager;
+  private UpsertConfig.Mode _upsertMode;
+  private List<String> _primaryKeyColumns;
+  private String _timeColumnName;
+
   public RealtimeTableDataManager(Semaphore segmentBuildSemaphore) {
     _segmentBuildSemaphore = segmentBuildSemaphore;
+    _upsertMetadataTableManager = new UpsertMetadataTableManager();
```

Comment:
This should only be initialized when upsert is enabled, or `addSegment()` will always pass the non-null UpsertMetadataTableManager, which will enable the upsert for the segments